Enable cloud-provider in GCE based deployments.

shashidharatd opened the pull request:
This PR is in continuation to migrate deployment mechanism used by federation to k8s-anywhere as listed out in https://github.com/kubernetes/test-infra/issues/3858. This is the 5th task in that series.

Federation CI jobs needs k8s clusters deployed with cloud-provider enabled for load-balancer type services and persistent volumes.

/cc @luxas @pipejakob @madhusudancs @kubernetes/sig-federation-pr-reviews

shashidharatd commented:
/assign @pipejakob

pipejakob commented:
Big apologies -- I've been out with a bad case of the flu for the past few days, but am catching up on things today.

shashidharatd commented:
No problem @pipejakob, take your time. Hope all is well now.

pipejakob commented on phase3/Kconfig:
```diff
@@ -38,6 +38,13 @@ config phase3.weave_net
+
+config phase3.gce_storage_class
+	depends on phase2.enable_cloud_provider
+	bool "Deploy default storage class?"
```
Is there a case where someone might want this enabled but not have the phase1 provider be gce? I'm wondering if this setting should also just depend on the phase1 provider option being set to gce.

pipejakob commented on phase1/gce/configure-vm-kubeadm.sh:
```diff
@@ -27,13 +30,38 @@ else
-    if [[ -n "$KUBERNETES_VERSION" ]]; then
-      OPTS="--kubernetes-version $KUBERNETES_VERSION"
+      cat <<EOF > kubeadm.conf
+kind: MasterConfiguration
```
As a heads-up, this section is being reworked as part of upgrade functionality being added here:

https://github.com/kubernetes/kubernetes-anywhere/pull/450/files#diff-990f027176aa91f2d9cdff0b6278c1bdR65

That change is to persist the flags used during `kubeadm init` so that we can reuse them later when upgrading a cluster. I assume that using the config file instead is really our only option to support settings like `controllerManagerExtraArgs`, but that is going to become another merge conflict soon.

luxas commented:
We'll just merge this after upgrade support is merged, and @shashidharatd can just put the new args in the config file as well

However, I really hope we don't have to do this, and instead can use the out-of-tree GCE cloud provider as an addon. cc @cheftako as your PR is needed for this to work

shashidharatd commented:
@luxas, @cheftako, i am curious to know when is the timeline we could use out-of-tree GCE cloud provider. federation need to start using the kbeadm based cluster deployment soon (like within a week or two). Should we wait until out-of-tree GCE cloud-provider is available or continue to use the current in-tree GCE cloud-provider?

Also @luxas, `--cloud-provider` flag is already removed from `kubeadm init`, so i cannot use the args passing method to `kubeadm` and instead should rely on config file to pass args to kubeadm. So i am in a  fix about what to do. Really appreciate any suggestions on way forward. Thanks !

shashidharatd commented:
@luxas, @cheftako, i am assuming external cloud-provider feature for GCE is not yet available and raised this PR to use the in-tree GCE cloud-provider and we shall update the code once it is available.

@jessicaochen i have modified the argument passing to `kubeadm init` from flags to config file. i have also updated the code path for upgrade scenario and require your review.

@pipejakob, i hope to have addressed your comments. request you to PTAL again.

Thanks !

luxas commented:
cc @jagosan

jessicaochen commented on phase2/kubeadm/do:
```diff
@@ -18,13 +18,14 @@ upgrade_master() {
   case "${UPGRADE_METHOD}" in
     "init")
-      execute_host ${MASTER} "sudo kubeadm init --skip-preflight-checks --kubernetes-version ${KUBERNETES_VERSION} ${INIT_PARAM}"
+      execute_host ${MASTER} "sudo kubeadm init --skip-preflight-checks --config ${KUBEADM_CONFIG_FILE}"
```
My understanding is this will not work. The init upgrade path is used for 1.6 -> 1.7. I believe init with config did not exist in 1.7. That is why I went with specifying params.

jessicaochen commented on phase1/gce/configure-vm-kubeadm.sh:
```diff
@@ -53,21 +56,50 @@ if [[ "${KUBEADM_VERSION}" != "${KUBELET_VERSION}" ]]; then
+EOF
+    fi
+
+    kubeadm init --skip-preflight-checks --config $KUBEADM_CONFIG_FILE
```
Will this work on a 1.7 cluster? Seems like it was introduced in 1.8.
https://github.com/kubernetes/kubernetes/commit/28117323bc92ebfc0eec8ade6531085e589b7fcd

jessicaochen commented on phase2/kubeadm/do:
```diff
@@ -18,13 +18,14 @@ upgrade_master() {
       ;;
     "upgrade")
-      execute_host ${MASTER} "sudo kubeadm config upload from-flags ${INIT_PARAM}"
+      execute_host ${MASTER} "sudo kubeadm config upload from-file --config ${KUBEADM_CONFIG_FILE}"
```
@luxas Is it needed to upload the config file if the config file was used with kubeadm init? 

If it is not needed, we can probably take out the logic uploading the file.

luxas replied:
It is needed if the cluster was v1.7.x before upgrade.

luxas commented:
`--config` has been around a very long time (~1.5), but the format of the file has changed slightly between versions. From what I saw, the args used there are all good (has been the same for many versions)

jessicaochen commented:
In that case /lgtm from an upgrade perspective.

shashidharatd commented:
@pipejakob, @luxas, request to have a look into this PR.
I have tested the installation scenario and it works well. But i have not tested the upgrade scenario. @jessicaochen request any suggestion or pointers to test the upgrade scenario.

Thanks !

shashidharatd commented:
/cc @marun, @irfanurrehman

shashidharatd commented:
I did the upgrade test by first deploying v1.7.8 cluster. Then i upgraded to v1.8.0.
Upgrade scenario is working along with the changes in this PR.

@pipejakob, can you please take it forward. Thanks !

pipejakob commented:
Taking a look now.

pipejakob commented:
/lgtm
